### PR TITLE
Ingela/tftp polish/otp 19114

### PR DESCRIPTION
**Pull request by IngelaAndin**
Polish tftp docs and remove legacy dependencies

### CODE REVIEW

File: lib/tftp/doc/guides/introduction.md [hunk 1]
```diff
@@ -17,30 +17,27 @@ limitations under the License.
...
-The only feature that not is implemented is the `netascii` transfer mode.
-
-## Prerequisites
-
-It is assumed that the reader is familiar with the Erlang programming language,
-concepts of OTP, and has a basic understanding of the TFTP protocol.
+# Overview
+
+TFTP is a very simple protocol used to transfer files over Transport
+datagram protcol UDP.  It is from this that its name comes, Trivial
```
Comment (RaimoNiskanen) [inline lib/tftp/doc/guides/introduction.md:23]: Awkward phrasing: "It is from this that its name comes, "
How about "Therefore its name: "

File: lib/tftp/doc/guides/introduction.md [hunk 2]
```diff
@@ -17,30 +17,27 @@ limitations under the License.
...
+
+`tftp` uses a callback module to handle the file transfer. Two such callback
+modules are provided, `tftp_binary` and `tftp_file`. See
+[read_file/3](`tftp:read_file/3`) and [write_file/3](`tftp:write_file/3`) for
+details. You can also implement your own callback modules, see
+[callbacks](`m:tftp#callbacks`).
+
+# Security Considerations
+
+As stated in RFC be aware that TFTP includes no login or access
```
Comment (RaimoNiskanen) [inline lib/tftp/doc/guides/introduction.md:38]: Which RFC (all RFC:s)?
"includes" - would that be "implements" or "has"

File: lib/tftp/doc/guides/introduction.md [hunk 3]
```diff
@@ -17,30 +17,27 @@ limitations under the License.
...
+control mechanisms, care must be taken in the rights granted to a TFTP
+server process so as not to violate the security of the server hosts
+file system.  TFTP is often installed with controls such that only
```
Comment (RaimoNiskanen) [inline lib/tftp/doc/guides/introduction.md:41]: "installed with controls" - would that be "installed with restrictions" or "installed in a restricted environment"?

File: lib/tftp/src/tftp.erl [hunk 4]
```diff
@@ -25,209 +25,127 @@
...
-%%%      is forwarded to the user and returned from read_file/3 or
-%%%      write_file/3.
-%%%-------------------------------------------------------------------
-
 -module(tftp).
+
+-moduledoc(#{titles =>
+                 [{function,<<"Client API">>},
+                  {function,<<"Server API">>},
+                  {function,<<"Client and Server API">>}]}
```
Comment (RaimoNiskanen) [inline lib/tftp/src/tftp.erl:33]: I cannot find "Client and Server API" in the built documentation

File: lib/tftp/src/tftp.erl [hunk 5]
```diff
@@ -25,209 +25,127 @@
...
+
+The only feature that not is implemented in this release is
+the "netascii" transfer mode.
+
+The start/1 function starts a daemon process which, listens for UDP
+packets on a port. When it receives a request for read or write it
+spawns a temporary server process which handles the actual transfer
+of the file. On the client side the read_file/3 and write_file/3
+functions spawns a temporary client process which establishes contact
+with a TFTP daemon and performs the actual transfer of the file.
```
Comment (RaimoNiskanen) [inline lib/tftp/src/tftp.erl:57]: "start/1", "read_file/3", and "write_file/3"; why not have links here with "code" typeface" like this: "`start/1`", ...

File: lib/tftp/src/tftp.erl [hunk 6]
```diff
@@ -25,209 +25,127 @@
...
+
+## Callbacks
+
+A `tftp` callback module is to be implemented as a `tftp` behavior and export
+the functions listed in the following.
+
+On the server side, the callback interaction starts with a call to `open/5` with
+the registered initial callback state. `open/5` is expected to open the
+(virtual) file. Then either function [`read/1`](`c:read/1`) or
+[`write/2`](`c:write/2`) is invoked repeatedly, once per transferred block. At
```
Comment (RaimoNiskanen) [inline lib/tftp/src/tftp.erl:70]: "[`read/1`](`c:read/1`)" and so on here would work just as well as simply "`c:read/1`".  In some documentation e.g. `gen_server` the convention [`Module:read/1`](`c:read/1`) is used, but apparently not in this documentation.

File: lib/tftp/src/tftp.erl [hunk 7]
```diff
@@ -25,209 +25,127 @@
...
+all option as-is or reject the transfer. Then the callback interaction follows
+the same pattern as described for the server side. When the last block is
+encountered in [`read/1`](`c:read/1`) or [`write/2`](`c:write/2`), the returned
+state is forwarded to the user and returned from `read_file`/3 or
+[`write_file/3`](`write_file/3`).
+
+If a callback (performing the file access in the TFTP server) takes too long
+time (more than the double TFTP time-out), the server aborts the connection and
+sends an error reply to the client. This implies that the server releases
+resources attached to the connection faster than before. The server simply
```
Comment (RaimoNiskanen) [inline lib/tftp/src/tftp.erl:93]: before what?

File: lib/tftp/src/tftp.erl [hunk 8]
```diff
@@ -243,158 +161,79 @@ of them differs a little. The available `option()`s are as follows:
...
+- **`{use_tsize, boolean()}`**
 
   Flag for automated use of option `tsize`. With this set to `true`, the
   [`write_file/3`](`write_file/3`) client determines the filesize and sends it
   to the server as the standardized `tsize` option. A
   [`read_file/3`](`read_file/3`) client acquires only a filesize from the server
   by sending a zero `tsize`.
 
-- **`{max_tsize, MaxTsize}`** - `MaxTsize = int() | infinity`
+- **`{max_tsize, MaxTsize::pos_integer() | infinity}`**
```
Comment (RaimoNiskanen) [inline lib/tftp/src/tftp.erl:182]: Here and in many places below `= int()` has been replaced with `:: pos_integer()`.  Is `0` really invalid in all these places, or should some be `:: non_neg_integer()`?

File: lib/tftp/src/tftp.erl [hunk 9]
```diff
@@ -243,158 +161,79 @@ of them differs a little. The available `option()`s are as follows:
...
-- **`{TftpKey, TftpVal}`** - `TftpKey = string()`  
-  `TftpVal = string()`
+- **TftpOption::option()**
 
   Name and value of a TFTP option.
 
-- **`{reject, Feature}`** - `Feature = Mode | TftpKey`  
-  ` Mode = read | write`  
-  ` TftpKey = string()`
+- **`{reject, Feature:: acess() | TftpKey::string()}`**
```
Comment (RaimoNiskanen) [inline lib/tftp/src/tftp.erl:196]: `acess` is spelled `access`

File: lib/tftp/src/tftp.erl [hunk 10]
```diff
@@ -590,7 +427,7 @@ matching regexp is found.
 -spec write_file(RemoteFilename, LocalFilename, Options) ->
         {ok, LastCallbackState} | {error, Reason} when
     RemoteFilename    :: file:filename(),
-    LocalFilename     :: file:filename_all(),
+    LocalFilename     :: file:filename_all() | binary,
```
Comment (RaimoNiskanen) [inline lib/tftp/src/tftp.erl:430]: Is the atom `binary` really correct here, or should that be `binary()`?

File: lib/tftp/doc/guides/introduction.md (hunk 1 above)
Comment (IngelaAndin) [reply lib/tftp/doc/guides/introduction.md:23]: Changed it to this:

Trivial File Transfer Protocol (TFTP) is a very simple protocol used
to transfer files over the transport datagram protocol UDP.

File: lib/tftp/doc/guides/introduction.md (hunk 2 above)
Comment (IngelaAndin) [reply lib/tftp/doc/guides/introduction.md:38]: It is a citation, I add citation marks and a link to RFC.

File: lib/tftp/doc/guides/introduction.md (hunk 3 above)
Comment (IngelaAndin) [reply lib/tftp/doc/guides/introduction.md:41]: Part of the citation.

File: lib/tftp/src/tftp.erl (hunk 5 above)
Comment (IngelaAndin) [reply lib/tftp/src/tftp.erl:57]: Agree that this would be an improvement, fixed.

File: lib/tftp/src/tftp.erl (hunk 7 above)
Comment (IngelaAndin) [reply lib/tftp/src/tftp.erl:93]: I have now idea just part of the original formulation, I will remove this sentence it does not add any value as is.

File: lib/tftp/src/tftp.erl (hunk 9 above)
Comment (IngelaAndin) [reply lib/tftp/src/tftp.erl:196]: fixed

File: lib/tftp/src/tftp.erl [hunk 11]
```diff
@@ -243,158 +161,79 @@ of them differs a little. The available `option()`s are as follows:
...
-  ` TftpKey = string()`
+- **`{reject, Feature:: acess() | TftpKey::string()}`**
 
   Controls which features to reject. This is mostly useful for the server as it
   can restrict the use of certain TFTP options or read/write access.
 
-- **`{callback, {RegExp, Module, State}}`** - `RegExp = string()`  
-  `Module = atom()`  
-  `State = term()`
+- **`{callback, {RegExp ::string(), Module::atom(), State :: term()}}`**
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/tftp/src/tftp.erl:201]: I would, good catch!

File: lib/tftp/src/tftp.erl (hunk 10 above)
Comment (IngelaAndin) [reply lib/tftp/src/tftp.erl:430]: Yes it is according to the code. file:filename_all  already covers binary().

**Review by RaimoNiskanen — APPROVED** [review]

**Pull request merged**